amethyst_tiles/src/pass.rs

```diff
@@ -0,0 +1,444 @@
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct DrawTiles2D<
+    B: Backend,
+    T: Tile,
+    E: CoordinateEncoder,
+    Z: DrawTiles2DBounds = DrawTiles2DBoundsDefault,
+> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    textures: TextureSub<B>,
+    vertex: DynamicVertexBuffer<B, TileArgs>,
+    sprites: OrderedOneLevelBatch<TextureId, TileArgs>,
+    change: util::ChangeDetection,
+
+    env: DynamicUniform<B, TileMapArgs>,
+
+    #[derivative(Debug = "ignore")]
+    _marker: PhantomData<(T, E, Z)>,
+}
+
+impl<B: Backend, T: Tile, E: CoordinateEncoder, Z: DrawTiles2DBounds> RenderGroup<B, World>
+    for DrawTiles2D<B, T, E, Z>
+{
+    #[allow(clippy::cast_precision_loss)]
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        world: &World,
+    ) -> PrepareResult {
+        #[cfg(feature = "profiler")]
+        profile_scope!("prepare");
+
+        let mut changed = false;
+        let (sprite_sheet_storage, tex_storage, hiddens, tile_maps, transforms) =
+            <(
+                Read<'_, AssetStorage<SpriteSheet>>,
+                Read<'_, AssetStorage<Texture>>,
+                ReadStorage<'_, Hidden>,
+                ReadStorage<'_, TileMap<T, E>>,
+                ReadStorage<'_, Transform>,
+            )>::fetch(world);
+
+        let sprites_ref = &mut self.sprites;
+        let textures_ref = &mut self.textures;
+
+        sprites_ref.swap_clear();
+
+        let CameraGatherer { projview, .. } = CameraGatherer::gather(world);
+
+        let mut tilemap_args = TileMapArgs {
+            proj: projview.proj,
+            view: projview.view,
+            map_coordinate_transform: Default::default(),
+            map_transform: Default::default(),
+            sprite_dimensions: Default::default(),
+        };
+
+        if let Some((tile_map, _, transform)) =
+            (&tile_maps, !&hiddens, transforms.maybe()).join().next()
+        {
+            let map_coordinate_transform: [[f32; 4]; 4] = (*tile_map.transform()).into();
+
+            let map_transform: [[f32; 4]; 4] = if let Some(transform) = transform {
+                (*transform.global_matrix()).into()
+            } else {
+                Matrix4::identity().into()
+            };
+
+            tilemap_args.map_coordinate_transform = map_coordinate_transform.into();
+            tilemap_args.map_transform = map_transform.into();
+            tilemap_args.sprite_dimensions = [
+                tile_map.tile_dimensions().x as f32,
+                tile_map.tile_dimensions().y as f32,
+            ]
+            .into();
+
+            let region = Z::bounds(tile_map, world);
+            region
+                .iter()
+                .filter_map(|coord| {
+                    let tile = tile_map.get(&coord).unwrap();
```

Comment:
This unwrap assumes the bounds are always precise. Also, there is no need for so much nesting.

Should just be 
```rust
let tile = tile_map.get(&coord)?;
let sprite_number = tile.sprite(coord, world)?;
```

Reply:
I don't want to softly fail for incorrect values, sprites or coordinates - I wanted to fail. Should I wrap those failures in `debug_assertions` gates?

Reply:
A sprite can be `None` for a tile, making it an empty tile.

Reply:
Empty tiles are exactly meant to return `None` here and be filtered out from rendering, so that checks out. Requiring that `bounds` call is always precise might be a stretch though. I can easily imagine a situation where i just use camera data to calculate that culling, I'd like to be not bothered about always checking the size of map as well. The renderer should calculate the region intersection before iterating, it's super low cost operation that improves developer experience.